### ISSUE TITLE
Sporadic crashes when Mapbox library is loaded: java.lang.UnsatisfiedLinkError at com.mapbox.common.loader.MapboxLibraryLoader.load

### ISSUE THREAD

**Issue by moritzzzzz** [mapbox/mapbox-maps-android#1109]
## Environment

- Android OS version: Android 11
- Devices affected: 83% Samsung, 17% Google
- Maps SDK Version:
```
implementation "com.mapbox.maps:android:10.1.0"
implementation "com.mapbox.mapboxsdk:mapbox-sdk-turf:5.8.0"
```
## Observed behavior and steps to reproduce
Sporadic crashes reported via crashlytics when loading Mapbox libraries.

## Expected behavior
No crashes.

## Additional links and references

**Stack trace**

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.installProvider (ActivityThread.java:7769)
  at android.app.ActivityThread.installContentProviders (ActivityThread.java:7309)
  at android.app.ActivityThread.handleMakeApplication (ActivityThread.java:7166)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:7134)
  at android.app.ActivityThread.access$1500 (ActivityThread.java:274)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2102)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:237)
  at android.app.ActivityThread.main (ActivityThread.java:8167)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:496)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1100)
Caused by: androidx.startup.StartupException: 
  at androidx.startup.AppInitializer.doInitialize (AppInitializer.java:162)
  at androidx.startup.AppInitializer.discoverAndInitialize (AppInitializer.java:198)
  at androidx.startup.InitializationProvider.onCreate (InitializationProvider.java:38)
  at android.content.ContentProvider.attachInfo (ContentProvider.java:2113)
  at android.content.ContentProvider.attachInfo (ContentProvider.java:2087)
  at android.app.ActivityThread.installProvider (ActivityThread.java:7764)
  at android.app.ActivityThread.installContentProviders (ActivityThread.java:7309)
  at android.app.ActivityThread.handleMakeApplication (ActivityThread.java:7166)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:7134)
  at android.app.ActivityThread.access$1500 (ActivityThread.java:274)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2102)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:237)
  at android.app.ActivityThread.main (ActivityThread.java:8167)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:496)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1100)
Caused by: java.lang.UnsatisfiedLinkError: 
  at java.lang.Runtime.loadLibrary0 (Runtime.java:1067)
  at java.lang.Runtime.loadLibrary0 (Runtime.java:1007)
  at java.lang.System.loadLibrary (System.java:1667)
  at com.mapbox.common.loader.MapboxLibraryLoader.load (MapboxLibraryLoader.kt:19)
  at com.mapbox.common.MapboxSDKCommonInitializer.create (MapboxSDKCommonInitializer.kt:62)
  at com.mapbox.common.MapboxSDKCommonInitializer.create (MapboxSDKCommonInitializer.kt:11)
  at androidx.startup.AppInitializer.doInitialize (AppInitializer.java:155)
  at androidx.startup.AppInitializer.discoverAndInitialize (AppInitializer.java:198)
  at androidx.startup.InitializationProvider.onCreate (InitializationProvider.java:38)
  at android.content.ContentProvider.attachInfo (ContentProvider.java:2113)
  at android.content.ContentProvider.attachInfo (ContentProvider.java:2087)
  at android.app.ActivityThread.installProvider (ActivityThread.java:7764)
  at android.app.ActivityThread.installContentProviders (ActivityThread.java:7309)
  at android.app.ActivityThread.handleMakeApplication (ActivityThread.java:7166)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:7134)
  at android.app.ActivityThread.access$1500 (ActivityThread.java:274)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2102)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:237)
  at android.app.ActivityThread.main (ActivityThread.java:8167)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:496)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1100)
```

**Comment by yunikkk** [comment]
Seems it's the same issue with library loading as https://github.com/mapbox/mapbox-gl-native-android/issues/416

**Comment by pf-burak-sinan** [comment]
We've started to get this exception lately as well. Here are the details, I hope it helps:
Stack trace:
```
java.lang.RuntimeException: Unable to get provider androidx.startup.InitializationProvider: androidx.startup.StartupException: androidx.startup.StartupException: androidx.startup.StartupException: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/package-name-VCfnYXvvcgeoxfVicOChTA==/base.apk"],nativeLibraryDirectories=[/data/app/package-name-VCfnYXvvcgeoxfVicOChTA==/lib/x86, /system/lib, /vendor/lib]]] couldn't find "libmapbox-common.so"
at android.app.ActivityThread.installProvider(ActivityThread.java:6242)
at android.app.ActivityThread.installContentProviders(ActivityThread.java:5805)
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5722)
at android.app.ActivityThread.-wrap1(Unknown Source:0)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1656)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6494)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/package-name-VCfnYXvvcgeoxfVicOChTA==/base.apk"],nativeLibraryDirectories=[/data/app/package-name-VCfnYXvvcgeoxfVicOChTA==/lib/x86, /system/lib, /vendor/lib]]] couldn't find "libmapbox-common.so"
at java.lang.Runtime.loadLibrary0(Runtime.java:1011)
at java.lang.System.loadLibrary(System.java:1657)
at com.mapbox.common.loader.MapboxLibraryLoader.load(MapboxLibraryLoader.kt:6)
at com.mapbox.common.CoreInitializer.create(CoreInitializer.kt:2)
at com.mapbox.common.CoreInitializer.create(CoreInitializer.kt:1)
at wj.b(AppInitializer.java:88)
... 17 more

```
So far we faced this issue on 2 devices repeatedly:
- Android 8.1.0; Nexus 5X on January 23
- Android 7.1.2; SM-T555 on January 18

Dependency details:
```
implementation("com.mapbox.maps:android:10.10.0") {
        exclude(group = "com.mapbox.plugin", module = "maps-compass")
        exclude(group = "com.mapbox.plugin", module = "maps-attribution")
        exclude(group = "com.mapbox.plugin", module = "maps-scalebar")
        exclude(group = "com.mapbox.plugin", module = "maps-overlay")
        exclude(group = "com.mapbox.plugin", module = "maps-locationcomponent")
        exclude(group = "com.mapbox.plugin", module = "maps-viewport")
    }
```

**Comment by norman-kapschefsky** [comment]
Also spiking for us now rapidly from today on:

```
Fatal Exception: java.lang.RuntimeException: Unable to get provider androidx.startup.InitializationProvider: androidx.startup.StartupException: androidx.startup.StartupException: java.lang.UnsatisfiedLinkError: dlopen failed: library "libmapbox-common.so" not found
       at android.app.ActivityThread.installProvider(ActivityThread.java:8215)
       at android.app.ActivityThread.installContentProviders(ActivityThread.java:7746)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7566)
       at android.app.ActivityThread.access$1500(ActivityThread.java:301)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2177)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8653)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

~80% of that crash are on Android 11 where 80% is Samsung Galaxy A12 atm. We use mapbox `10.6.0`.

**Comment by a-h-mzd** [comment]
We're getting the same error and we use mapbox `10.10.0`.

**Comment by PavlosTze** [comment]
Still an issue at `10.11.0`.

**Comment by alingocab** [comment]
Hi! I'm getting this error more and more often. Yesterday I pushed an update to users and it started to happen even more often than before. Please see the image.

![Screenshot 2023-02-16 at 09 24 37](https://user-images.githubusercontent.com/91056746/219296209-2669c204-a926-4865-8661-39e08468c918.jpeg)

![Screenshot 2023-02-16 at 09 24 32](https://user-images.githubusercontent.com/91056746/219296191-8f94c1a0-1c65-409e-b0ca-22de66e0d5d9.jpeg)

![Screenshot 2023-02-16](https://user-images.githubusercontent.com/91056746/219295996-37f832b4-6fe6-4bdd-a50a-b940b6bcdf46.jpeg)

**Comment by alingocab** [comment]
BTW, I'm using this Mapbox dependencies:
com.mapbox.mapboxsdk:mapbox-sdk-services:6.10.0
com.mapbox.maps:android:10.10.1

**Comment by jush** [comment]
@alingocab Have you released an app with [crashlytics 18.3.0](https://firebase.google.com/support/release-notes/android#crashlytics_v18-3-0) (or higher) recently?

**Comment by alingocab** [comment]
@jush I'm using com.google.firebase:firebase-bom:31.2.1. As I can see, firebase-bom:31.2.1 is using crashlytics 18.3.5.

**Comment by jush** [comment]
Thanks, [Crashlytics release 18.3.0](https://firebase.google.com/support/release-notes/android#crashlytics_v18-3-0) introduced this change that might explain why you see it more often:
> Improved crash reporting reliability for crashes that occur early in the app's lifecycle

**Comment by alingocab** [comment]
Ok. It is very good that I can see it, but the main issue is that the app is actually crashing with this error. I can see a huge number of crashes generated by some of the devices like Nexus 5x, Samsung Galaxy S9+ etc. 

<img width="1321" alt="Screenshot 2023-02-17 at 10 13 17" src="https://user-images.githubusercontent.com/91056746/219589515-985c5d42-33a9-4af3-ac88-14bc6299ff71.png">

**Comment by alingocab** [comment]
I will attach full crash details from Firebase.

[mapbox_issue_064183bd8c8ab9d3bbdbadb2f179601c_crash_session_63EF375003D9000101776F7ACB04C50F_DNE_0_v2_stacktrace.txt](https://github.com/mapbox/mapbox-maps-android/files/10764692/mapbox_issue_064183bd8c8ab9d3bbdbadb2f179601c_crash_session_63EF375003D9000101776F7ACB04C50F_DNE_0_v2_stacktrace.txt)

**Comment by michaelEllisUy** [comment]
Same issue here as well. Crash log looks just like the one posted by alingocab.

**Comment by boiler23** [comment]
I also observe these crashes after migrating from v9 to v10. 
<img width="415" src="https://user-images.githubusercontent.com/215357/222895951-ae1fa908-698e-4c1d-8d25-a667080b7b9d.png">

**Comment by alingocab** [comment]
Hello. Yesterday we updated com.mapbox.maps:android to 10.11.1. Still, those crashes started to flow immediately after update. Please see the screenshot attached. It would be very helpful if someone from Mapbox team can provide a solution to this. This thread has been started more than 1 year ago and still no practical response. Thanks in advance.
<img width="1522" alt="Screenshot 2023-03-09 at 11 16 11" src="https://user-images.githubusercontent.com/91056746/223975918-c3f3cffd-5d3b-41ee-bcd2-5f84748fb9dc.png">

**Comment by LeeQiuuu** [comment]
i have the same problem

**Comment by nwilsongme** [comment]
Same here.  Has there been any resolution?

**Comment by alingocab** [comment]
Update mabox sdk to 10.12.0 today, issue still present, a lot of crashes coming up. Seems like this crash is not important for mapbox team...

**Comment by matthew-niemann** [comment]
Also seeing this issue since using Crashlytics 18.3.0+, but I have not physically seen the crash myself, nor have I received a "startup crash" report from our beta testers or Play Store users. Perhaps it is either a 'silent' crash, or happens only once per user and they assume nothing of it?

**Comment by iamwent** [comment]
I have the same crash with device Nexus 5X 8.1.0 while Mapbox Android is 10.12.0
``

**Comment by MrTheGood** [comment]
> I have the same crash with device Nexus 5X 8.1.0 while Mapbox Android is 10.12.0

For me this is the same: while there are only 3 users having this crash, they're all on a Nexus 5x with Android 8.1.0

**Comment by alingocab** [comment]
> > I have the same crash with device Nexus 5X 8.1.0 while Mapbox Android is 10.12.0
> 
> For me this is the same: while there are only 3 users having this crash, they're all on a Nexus 5x with Android 8.1.0

The reason you see Nexus 5x is because that's a device used by Google to run some tests on your app before pushing it to Google Play. In my case, beside this Nexus 5x I see a lot of other devices owned by real users.

**Comment by iamwent** [comment]
> 

It's interesting. But there are also some crashes on the Galaxy devices.

**Comment by kiryldz** [comment]
Could you please check if [following](https://github.com/mapbox/mapbox-maps-android/issues/1173#issuecomment-1062707575) helps?

**Comment by boiler23** [comment]
In my case I don't remove any initialization providers, so the solution above won't help I believe. Merged manifest file also looks fine IMO:
```
<provider
    android:name="androidx.startup.InitializationProvider"
    android:authorities="com.myapp.androidx-startup"
    android:exported="false" >
    <meta-data
        android:name="com.mapbox.maps.loader.MapboxMapsInitializer"
        android:value="androidx.startup" />
    <!-- This entry makes MapboxSDKCommonInitializer discoverable. -->
    <meta-data
        android:name="com.mapbox.common.MapboxSDKCommonInitializer"
        android:value="androidx.startup" />
    <meta-data
        android:name="androidx.emoji2.text.EmojiCompatInitializer"
        android:value="androidx.startup" />
    <meta-data
        android:name="androidx.lifecycle.ProcessLifecycleInitializer"
        android:value="androidx.startup" />
</provider>
```
I still didn't receive any complaints from users, and most of the cases happen in the background:
<img width="430" alt="screenshot" src="https://user-images.githubusercontent.com/215357/231543002-668f6ae0-ed00-479d-9400-ba6ffbd1df44.png">

Maybe this will give you some clues for investigation: I have two separate processes in the app. So the hypothesis is this initialization issue happens in the second process, which doesn't interact with Mapbox and doesn't provide any UI (and that's probably why there are no complaints from users).

**Comment by kawan2019** [comment]
Still an issue at 10.12.2
<img width="1299" alt="Screen Shot 2023-04-23 at 12 14 36 PM" src="https://user-images.githubusercontent.com/57112131/233830923-49f274a8-e41b-4f07-9d6a-a44892e11841.png">

**Comment by leanh128** [comment]
packagingOptions {
  pickFirst '/armeabi-v7a/libmapbox-gl.so'
  pickFirst '/x86/libmapbox-gl.so'
  pickFirst '/arm64-v8a/libmapbox-gl.so'
  pickFirst '/x86_64/libmapbox-gl.so'
  pickFirst '**/x86/libmapbox-gl.so'
}

**Comment by tasomaniac** [comment]
> Could you please check if [following](https://github.com/mapbox/mapbox-maps-android/issues/1173#issuecomment-1062707575) helps?

The above workaround initially made a lot of sense to us but unfortunately it did not work for us. After applying this, the stack trace has changed. Now we can see that the initialization is explicitly called within our `Application` class but immediately after that the app crashes with the same reason.

**Comment by boiler23** [comment]
Unfortunately, solution from @leanh128 didn't help me either. Currently using 10.14.0-beta.1.

**Comment by iamwent** [comment]
> Unfortunately, solution from @leanh128 didn't help me either. Currently using 10.14.0-beta.1.

I also tried this solution, failed 😟

**Comment by PavlosTze** [comment]
@kiryldz @ank27 @jush @alexander-kulikovskii This is still happening on 10.15.1. Any news on that? There have been hundreds of crashes.

**Comment by arthurljones** [comment]
We also have about 100 crashes from this in the last 7 days, and 1900 in the last 90 days. A fix or functioning workaround would be greatly appreciated.

**Comment by Ranafugaz** [comment]
same problem here

**Comment by AndroidPat** [comment]
This issue generates hundreds of crashes for us as well. We use:

        
        <provider
            android:name="androidx.startup.InitializationProvider"
            android:authorities="${applicationId}.androidx-startup"
            android:exported="false"
            tools:node="merge">
            <meta-data
                android:name="androidx.work.WorkManagerInitializer"
                android:value="androidx.startup"
                tools:node="remove" />
        </provider>

but reading the comments above adding AppInitializer.getInstance(this)        .initializeComponent(MapboxMapsInitializer::class.java) is not going to fix this issue.

**Comment by paul-wag** [comment]
Has there been any comments or updates from Mapbox on this?

**Comment by FrancisDx** [comment]
Come on, any update? This issue is the biggest source of crashes in my apps (by a lot)

**Comment by PavlosTze** [comment]
Same here, seems Mapbox team doesn't really care... We are waiting for updates for months. @kiryldz @ank27 @jush @alexander-kulikovskii

**Comment by boiler23** [comment]
So, seems I found a dirty workaround to this issue. At least, it no longer pops up in the new app version (before, it was a top-1 crasher).

**TLDR**: I don't initialize Mapbox on application startup anymore.

**Longer answer**:
My use case might be a bit specific, as users can use other map providers in the app, besides Mapbox.
Moreover, I noticed that amount of these crashes is comparable with the number of Mapbox users in the app. If every second-third user with Mapbox would see this issue - there should be a lot of user complaints, right? However, that wasn't the case. So, I assumed that _the users who aren't interacting with Mapbox in the app are probably the ones suffering from this issue_.
The obvious solution that came to my mind - is to postpone Mapbox initialization until the user explicitly interacts with it (e.g. fragment with Mapbox maps is created).
However, the Mapbox SDK does initialize Mapbox maps on application start (via Jetpack Startup) and isn't designed for such lazy initialization.
My initial approach was to remove the jetpack startup code, related to Mapbox from the manifest, and do init it inside the `Fragment` code:
```
      <provider
          android:name="androidx.startup.InitializationProvider"
          android:authorities="${applicationId}.androidx-startup"
          android:exported="false"
          tools:node="merge">
          <meta-data
              android:name="com.mapbox.maps.loader.MapboxMapsInitializer"
              android:value="androidx.startup"
              tools:node="remove" />
          <meta-data
              android:name="com.mapbox.common.MapboxSDKCommonInitializer"
              android:value="androidx.startup"
              tools:node="remove" />
      </provider>
```
and then in fragment's `onCreate` (**not in the application's one!**):
```
AppInitializer.getInstance(appContext).initializeComponent(MapboxMapsInitializer::class.java)
```
This approach works sometimes, but sometimes it leads to another crash - some internal Mapbox code tries to access `lateinit var context`, which I believe is assigned during the application start. I examined the source code around it a bit better, and it turned out to be a broadcast receiver, `com.mapbox.common.location.LocationUpdatesReceiver` - it is most probably related to gathering telemetry by the Mapbox's engine. Now the problem became clear - broadcast receiver could be live even if the map fragment is not visible. The only solution to overcome this that came to my mind was to remove this broadcast receiver from the manifest, and launch it manually in `onCreateView`/`onDestroyView` of the fragment:
```
        <receiver
            android:name="com.mapbox.common.location.LocationUpdatesReceiver"
            tools:node="remove" >
        </receiver>
```
And, in the map fragment:
```
class MyMapFragment : Fragment() {
    ...
    private val locationUpdatesReceiver = LocationUpdatesReceiver()
    ...
    fun onCreateView(...) {
        ....
        appContext.registerReceiver(
               locationUpdatesReceiver,
               IntentFilter(LocationUpdatesReceiver.ACTION_PROCESS_LOCATION_UPDATES),
               Context.RECEIVER_NOT_EXPORTED,
         )
    }

    fun onDestroyView(...) {
        try {
            appContext.unregisterReceiver(locationUpdatesReceiver)
        } catch (e: IllegalArgumentException) {
            ...
        }
    }
}
```
Unregistering of the receiver is wrapped in the try/catch block, to prevent crashes for cases when the receiver is not active anymore.

This approach works for me, and I don't see any UnsatisfiedLinkError crashes with the new app version so far (out of ~200k installs).

I hope this will help folks from Mapbox to identify and fix the issue (so I can remove all these hacks from the codebase!), as well as other developers suffering from it to workaround this :)

**Comment by tasomaniac** [comment]
We tried removing the auto initializer and manually initializing. That reduced the crashes but they're still too many.

Great find about the BroadcastReceiver. If the SDK is not initialized yet, I am not sure if that receiver should be listening. That may even be a privacy issue.

This is hacky in the sense that you have some extra code that is not needed if auto initializer worked as expected. But actually it is not that hacky since it can be isolated in a non disruptive way.

**Comment by boiler23** [comment]
Yes, and I also consider it hacky in the sense that it uses some internals to the Mapbox SDK, so updating to a newer SDK version can become an issue in the future.

**Comment by tasomaniac** [comment]
That's a good point

**Comment by jush** [comment]
The Mapbox team has prepared a snapshot with a potential solution for the sporadic crash in the Maps SDK for Android library.

We have been investigating the sporadic crash in the Mapbox library and we have prepared a snapshot with a potential solution for you to test.

While we are confident that this snapshot will help to resolve the issue, we have not been able to reproduce it in our environments, therefore we cannot guarantee that it will be a complete fix. We encourage you to test the snapshot thoroughly and report any issues you find on this issue tracker: https://github.com/mapbox/mapbox-maps-android/issues/1109.

Your feedback will help us determine the effectiveness of the snapshot and identify any remaining issues.

Instructions for testing the snapshot:

1. Your app must not be using any other Mapbox SDK (e.g. [Navigation](https://docs.mapbox.com/android/navigation/guides/), [Search](https://docs.mapbox.com/android/search/guides/)).
2. Install the snapshot maven repository into your project by following the instructions at [Use snapshot releases in your app](https://github.com/mapbox/mapbox-maps-android/blob/lts/v10/DEVELOPING.md#use-snapshot-releases-in-your-app).
3. Replace your Mapbox SDK for Android version with version `10.16.2-SNAPSHOT.1115T1113Z.912f916`.
4. Run your app and test the functionality.
5. Publish your app to your users.
6. Report any issues you find on the issue tracker: https://github.com/mapbox/mapbox-maps-android/issues/1109.
7. We appreciate your cooperation as we work to resolve this issue. Thank you for your patience and understanding.

**Comment by paul-wag** [comment]
@jush thank you for providing the update. Just wanted to let you know we have an app that only uses search/autocomplete and the issue happens to us. It's one of our top crashes. 

> Your app must not be using any other Mapbox SDK (e.g. [Navigation](https://docs.mapbox.com/android/navigation/guides/), [Search](https://docs.mapbox.com/android/search/guides/)).

**Comment by tasomaniac** [comment]
Thanks for the update. We do use the Navigation dependency. So I guess we will have to wait until this makes it to the stable release

**Comment by guitabalipa** [comment]
@jush Do you have an estimate date for this fix to hit the stable version?

**Comment by AndroidPat** [comment]
> The Mapbox team has prepared a snapshot with a potential solution for the sporadic crash in the Maps SDK for Android library.
> 
> We have been investigating the sporadic crash in the Mapbox library and we have prepared a snapshot with a potential solution for you to test.
> 
> While we are confident that this snapshot will help to resolve the issue, we have not been able to reproduce it in our environments, therefore we cannot guarantee that it will be a complete fix. We encourage you to test the snapshot thoroughly and report any issues you find on this issue tracker: #1109.
> 
> Your feedback will help us determine the effectiveness of the snapshot and identify any remaining issues.
> 
> Instructions for testing the snapshot:
> 
> 1. Your app must not be using any other Mapbox SDK (e.g. [Navigation](https://docs.mapbox.com/android/navigation/guides/), [Search](https://docs.mapbox.com/android/search/guides/)).
> 2. Install the snapshot maven repository into your project by following the instructions at [Use snapshot releases in your app](https://github.com/mapbox/mapbox-maps-android/blob/lts/v10/DEVELOPING.md#use-snapshot-releases-in-your-app).
> 3. Replace your Mapbox SDK for Android version with version `10.16.2-SNAPSHOT.1115T1113Z.912f916`.
> 4. Run your app and test the functionality.
> 5. Publish your app to your users.
> 6. Report any issues you find on the issue tracker: [Sporadic crashes when Mapbox library is loaded: java.lang.UnsatisfiedLinkError at com.mapbox.common.loader.MapboxLibraryLoader.load #1109](https://github.com/mapbox/mapbox-maps-android/issues/1109).
> 7. We appreciate your cooperation as we work to resolve this issue. Thank you for your patience and understanding.

Thanks for trying to resolve it. It's a #1 crash in our app. I've tried pulling 10.16.2-SNAPSHOT.1115T1113Z.912f916 but I get 403 forbidden error. The version 10.16.2 pulls fine.

**Comment by kiryldz** [comment]
@AndroidPat 
Please make sure you have added the additional Maven mirror as described in the point 2 in the [instructions](https://github.com/mapbox/mapbox-maps-android/issues/1109#issuecomment-1812636156).

**Comment by AndroidPat** [comment]
> @AndroidPat Please make sure you have added the additional Maven mirror as described in the point 2 in the [instructions](https://github.com/mapbox/mapbox-maps-android/issues/1109#issuecomment-1812636156).

Thank you, I missed this bit!
I've changed the URL to the snapshot one and the error changed from https://mvnrepository.com/artifact/com.github.gundy/semver4j/com/mapbox/plugin/maps-lifecycle/10.16.2-SNAPSHOT.1115T1113Z.912f916/maps-lifecycle-10.16.2-SNAPSHOT.1115T1113Z.912f916.pom

to

Could not HEAD 'https://mvnrepository.com/artifact/com.github.gundy/semver4j/com/mapbox/base/common/0.8.0/common-0.8.0.pom'. Received status code 403 from server: Forbidden

**Comment by kiryldz** [comment]
> > @AndroidPat Please make sure you have added the additional Maven mirror as described in the point 2 in the [instructions](https://github.com/mapbox/mapbox-maps-android/issues/1109#issuecomment-1812636156).
> 
> Thank you, I missed this bit! I've changed the URL to the snapshot one and the error changed from https://mvnrepository.com/artifact/com.github.gundy/semver4j/com/mapbox/plugin/maps-lifecycle/10.16.2-SNAPSHOT.1115T1113Z.912f916/maps-lifecycle-10.16.2-SNAPSHOT.1115T1113Z.912f916.pom
> 
> to
> 
> Could not HEAD 'https://mvnrepository.com/artifact/com.github.gundy/semver4j/com/mapbox/base/common/0.8.0/common-0.8.0.pom'. Received status code 403 from server: Forbidden

@AndroidPat please _add_ the new mirror and do not replace it. Your gradle file should look somehow like:
```
allprojects {
    repositories {
        // your staff
        maven {
            url 'https://api.mapbox.com/downloads/v2/releases/maven'
            authentication {
                basic(BasicAuthentication)
            }
            credentials {
                username = "mapbox"
                password = YOUR_SECRET_TOKEN
            }
        }
        maven {
            url 'https://api.mapbox.com/downloads/v2/snapshots/maven'
            authentication {
                basic(BasicAuthentication)
            }
            credentials {
                username = "mapbox"
                password = YOUR_SECRET_TOKEN
            }
        }
    }
}
```

**Comment by AndroidPat** [comment]
https://gist.github.com/AndroidPat/173e70244bf05c386745bdd70398ec9c
No luck, I've tried changing the order and invalidating caches etc.
If you have any ideas, I would appreciate a comment on the gist, to keep this thread clean.

**Comment by oguzcelik** [comment]
Hey folks 🖖 We are planning to test the new 10.16.3 version to resolve the crash. But we are using the nav sdk for some of our applications https://github.com/mapbox/mapbox-navigation-android/releases/tag/v2.17.2

In the [release notes for the version](https://github.com/mapbox/mapbox-maps-android/releases/tag/v10.16.3) it mentions that if we use nav sdk, we might still face the crash. And in the nav sdk release notes it mentions the version is tested with an earlier version.

I have 2 questions to clear some confusions we have. 
- Does it mean that we will continue to have the crash that we discussed in the thread, or could there be more incompatibility with the nav sdk version we use and the fix in 10.16.3 ? 
- Do you have a plan/estimate on when the nav sdk would have this fix as well ? 

Thank you for providing a fix for this issue. It's been a long time that we had this as a top crash 🚀

**Comment by kiryldz** [comment]
Hey @oguzcelik!

> Does it mean that we will continue to have the crash that we discussed in the thread, or could there be more incompatibility with the nav sdk version we use and the fix in 10.16.3 ?

You will most likely see another crash saying that another (navigation related and not common described in this issue) `.so` native library could not be loaded. Also noting that using [Nav SDK 2.17.2](https://github.com/mapbox/mapbox-navigation-android/releases/tag/v2.17.2) with Maps 10.16.3 will most likely crash in runtime due to different native symbols. To be on the safe side Nav SDK should release another patch that's compatible with Maps 10.16.3.

> Do you have a plan/estimate on when the nav sdk would have this fix as well ?

No clear estimations for now sadly. As soon as we have details we will share them here. The high level plan is to actually have the fix in Maps v11 first thing and then focus on updating Nav / Search SDKs.

**Comment by felixkrautschuk** [comment]
I just tred out the latest version com.mapbox.maps:android:11.2.0-beta.1
but the app is still crashing on start:

> Caused by: com.mapbox.common.MapboxInitializerException: Failed to initialize: Attempt=2, exception=[StartupException], initializer called 3686 ms ago, exception.message=[androidx.startup.StartupException: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/data/app/~~QOSaTqLzwRmrFoH4QDBnDw==/de.maxity.mapbox.demo-7HnSklBqHWjkhF9tDLs1Xw==/base.apk!/lib/x86_64/libmapbox-common.so"...], exception.cause=[StartupException], exception.cause.message=[java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/data/app/~~QOSaTqLzwRmrFoH4QDBnDw==/de.maxity.mapbox.demo-7HnSklBqHWjkhF9tDLs1Xw==/base.apk!/lib/x86_64/libmapbox-common.so"...], extractedNativeLibs=[], isInstantApp=[false], isMainThread=[true]

**Comment by jush** [comment]
@felixkrautschuk Hi, this seems slightly different issue. This is not that the native library  is completely missing in the device but rather that `__emutls_get_address` symbol can't be found.

Could you please share more details about:
1. share detailed steps how to reproduce? if possible also repro project.
2. Your setup, what dependencies do you have in the project?
3. Device details where the issue is happening (model, OS version,...)
Thanks!

**Comment by felixkrautschuk** [comment]
@jush thanks for your response.

I created a new ticket here: https://github.com/mapbox/mapbox-maps-android/issues/2271

**Comment by kiryldz** [comment]
Hey all, Mapbox has implemented better way of initialization and it's available both for Maps v10 / v11 and Nav v2 / v3.

Here's the list of actual versions where this is implemented:
1. [Nav 2.19.0](https://github.com/mapbox/mapbox-navigation-android/releases/tag/v2.19.0) that uses [Maps v10.16.6](https://github.com/mapbox/mapbox-maps-android/releases/tag/v10.16.6).
2. [Core Framework v1.0.0-rc.3](https://docs.mapbox.com/android/navigation/api/coreframework/1.0.0-rc.3/) that uses [Maps v11.2.0](https://github.com/mapbox/mapbox-maps-android/releases/tag/v11.2.0).

### What does the fix cover

This means that exception described in this ticket happening on startup (as we're using AndroidX [initializers](https://developer.android.com/topic/libraries/app-startup) in SDK) should not happen anymore. **However** we did not find the actual root cause **why** the native libraries might be missing on startup, it seems to be the Android App Bundle implementation which is out of our control.

So the initialization workflow now works the following way:
1. We initialize as we did before using AndroidX [initializers](https://developer.android.com/topic/libraries/app-startup).
2. If initialization attempt on app startup failed we gracefully switch to "lazy" initialization. This means that Mapbox will initialize only when any actual Mapbox API will be called. We assume that the most common use-case is actually creating / inflating the `MapView` somewhere in activity / fragment so in that example it means that _if_ we failed to load native libraries on startup, the SDK will handle this for you and try to run all the initialization code right before creating the `MapView`.

Summing up: the `UnsatisfiedLinkError` should be eliminated without any changes in client code. The similar approach was proved to be working at least by @boiler23 in [this](https://github.com/mapbox/mapbox-maps-android/issues/1109#issuecomment-1806776527) comment so we hope it should help other users as well. P.S if you had https://github.com/mapbox/mapbox-maps-android/issues/1109#issuecomment-1806776527 workaround implemented - it should be reverted when switching to the Maps / Nav versions that contain the fix.

### What does the fix not cover 

As said above we did not find the actual reason why the native libraries might not be in place. So it's absolutely possible that even if we gracefully catch the exception for you on startup and do not crash - the libraries might still be not there when you first access the Mapbox APIs. The most obvious use-case for that is accessing some Mapbox APIs in your `Application.onCreate` which is actually triggered shortly after AndroidX initializers are run and there's high chance that the native libraries are still not there.

If the native libraries could not be loaded right before Mapbox function - our SDK could not work. However with that new approach it's still possible to recover in this situation but that will require adding code in your end applications based on the business logic you want to apply.

Example: You have an app that uses App Bundle and even after upgrading to latest version with a fix you see crashes (hopefully reduced amount). Looking into your code you realize that the `MapView` is shown right after entering your application.

```
// somewhere in your Activity.onCreate
val mapView = MapView(this)
```

In that case you could apply the following approach:

```
try {
  // if Mapbox was actually init successfully before - this call is no-op and cancels immediately;
  // if Mapbox was not init due to missing native libraries - we will try to initialize again
  // this thing could throw an exception if native libraries are not found
  BaseMapboxInitializer.init(MapboxSDKCommonInitializerImpl.class)
  // if we are hitting this line we could safely assume that initialization was fine and we could proceed
  val mapView = MapView(this)
} catch (e: Throwable) {
   // for example you could either some UI with `Retry` button and clicking that button runs this code again;
   // or you could ask the user try re-opening the application
   // or do something else based no your business needs
}
```

### Conclusion

We hope that the fix will work well for you. I am closing this ticket (as it's already way too huge) but please feel free to open new issues (also note that we greatly enhanced the underlying exception so it contains additional data that might help us to proceed even further hopefully in fully eliminating this in SDK). As an example there was already an [issue](https://github.com/mapbox/mapbox-maps-android/issues/2302) related to new approach and connected with accessing Mapbox API inside user's initializers.

**Comment by PavlosTze** [comment]
@kiryldz But we cannot upgrade to Mapbox [because of this](https://github.com/mapbox/mapbox-search-android/issues/171).

**Comment by dudeuter** [comment]
@PavlosTze the issue you shared is different from the one in this thread. I added some notes on your issue for anyone who's poking around.

This issue is for `java.lang.UnsatisfiedLinkError at com.mapbox.common.loader.MapboxLibraryLoader.load` which is related to the SDK's initializers being run.

For anyone getting the error from this thread the best trouble shooting steps will be:

1. Look in your build intermediates folder to make sure your merged AndroidManifest has entries for the Mapbox Initializers.
2. Make sure that your application is packaging the shared objects as expected

Some useful shell commands, run from your project directory.

`find ./app/build -type f -name "*.so"`

You should see some mapbox shared objects in your build intermediates. Additionally you can unpack your applications APK file to look for the shared objects as well. 

`grep --binary-files=without-match -Rin "initializer" ./app/build`

You should see an entry for a mapbox initializer; if it's omitted then there's something about your app's android manifest that's overriding the SDK's manifest to run the initializer; which loads the shared objects.

**Comment by dudeuter** [comment]
@PavlosTze can you confirm that you have the initializer as well? Lastly you can repeat the step above by unpacking your apk as well.

If everything there still exists then it's likely a conflict with how you have Multidexing configured; which is more of an Android issue that a Mapbox issue.

**Comment by PavlosTze** [comment]
@dudeuter I confirm that we have the initializer and we have the libraries in the APK as well, everything is there.

We don't have Multidexing configured, and these crashes did **not** take place for years until since the version mentioned in this issue.

**Comment by narko** [comment]
The issue seems to be still happening with the following versions:
```
mapboxSearch = "2.2.0"
mapboxMaps = "11.3.0"
mapboxNavigation = "3.0.1"
```

Any chance this could be reopened?

**Comment by kiryldz** [comment]
@narko please refer to `What this fix does not cover` section [here](https://github.com/mapbox/mapbox-maps-android/issues/1109#issuecomment-2017733523). If you have more information (e.g. reproducible code causing the crash) - please open the new issue with detailed information.

**Comment by narko** [comment]
@kiryldz unfortunately I have no clear evidence of why this is also affecting our code. I'll share the traces in case it gives someone a hint:

```
Fatal Exception: java.lang.RuntimeException
Unable to get provider androidx.startup.InitializationProvider: K3.c: K3.c: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "XXX"],nativeLibraryDirectories=[/data/app/XXX==/lib/x86, /system/lib, /vendor/lib]]] couldn't find "libSearchCore.so"

at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:187)
       at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:238)
       at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:206)
       at androidx.startup.InitializationProvider.onCreate(InitializationProvider.java:45)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:1917)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:1892)
       at android.app.ActivityThread.installProvider(ActivityThread.java:6239)
       at android.app.ActivityThread.installContentProviders(ActivityThread.java:5805)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5722)
       at android.app.ActivityThread.-wrap1()
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1656)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:164)
       at android.app.ActivityThread.main(ActivityThread.java:6494)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```


In our `Application.onCreate()` we use the following initializer:

```
AppInitializer.getInstance(this)
            .initializeComponent(MapboxMapsInitializer::class.java)
```
